crypto/src/utils.rs

```diff
@@ -0,0 +1,214 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// spell-checker:ignore cipherparams ciphertext
+
+use super::{gen_keypair, gen_keypair_from_seed, PublicKey, SecretKey, Seed, SEED_LENGTH};
+use hex_buffer_serde::Hex;
+use pwbox::{sodium::Sodium, ErasedPwBox, Eraser, Suite};
+use rand::thread_rng;
+use std::borrow::Cow;
+#[cfg(unix)]
+use std::os::unix::fs::{MetadataExt, OpenOptionsExt};
+use std::{
+    fs::{File, OpenOptions},
+    io::{Error, ErrorKind, Read, Write},
+    path::Path,
+};
+use toml;
+
+/// Creates a TOML file that contains encrypted `SecretKey` and returns `PublicKey` for the secret key.
+pub fn generate_keys_file<P: AsRef<Path>, W: AsRef<[u8]>>(
+    path: P,
+    pass_phrase: W,
+) -> Result<PublicKey, Error> {
+    let (pk, sk) = gen_keypair();
+    let keys = EncryptedKeys::encrypt(pk, &sk, pass_phrase)?;
+    let file_content =
+        toml::to_string_pretty(&keys).map_err(|e| Error::new(ErrorKind::Other, e))?;
+    let mut open_options = OpenOptions::new();
+    open_options.create(true).write(true);
+    #[cfg(unix)]
+    open_options.mode(0o600);
+    let mut file = open_options.open(path.as_ref())?;
+    file.write_all(file_content.as_bytes())?;
+
+    Ok(pk)
+}
+
+/// Reads and returns `PublicKey` and `SecretKey` from encrypted file located by path and returns its.
+pub fn read_keys_from_file<P: AsRef<Path>, W: AsRef<[u8]>>(
+    path: P,
+    pass_phrase: W,
+) -> Result<(PublicKey, SecretKey), Error> {
+    let mut key_file = File::open(path)?;
+
+    #[cfg(unix)]
+    validate_file_mode(key_file.metadata()?.mode())?;
+
+    let mut file_content = vec![];
+    key_file.read_to_end(&mut file_content)?;
+    let keys: EncryptedKeys =
+        toml::from_slice(file_content.as_slice()).map_err(|e| Error::new(ErrorKind::Other, e))?;
+    keys.decrypt(pass_phrase)
+}
+
+#[cfg(unix)]
+#[cfg_attr(feature = "cargo-clippy", allow(clippy::verbose_bit_mask))]
+fn validate_file_mode(mode: u32) -> Result<(), Error> {
+    if (mode & 0o077) == 0 {
+        Ok(())
+    } else {
+        Err(Error::new(ErrorKind::Other, "Wrong file's mode"))
+    }
+}
+
+struct PublicKeyHex;
+
+impl Hex<PublicKey> for PublicKeyHex {
+    fn create_bytes(value: &PublicKey) -> Cow<[u8]> {
+        Cow::Borrowed(&*value.as_ref())
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Result<PublicKey, String> {
+        PublicKey::from_slice(bytes)
+            .ok_or_else(|| "Couldn't create PublicKey from slice".to_owned())
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct EncryptedKeys {
+    #[serde(with = "PublicKeyHex")]
+    public_key: PublicKey,
+    secret_key: ErasedPwBox,
+}
+
+impl EncryptedKeys {
+    fn encrypt(
+        public_key: PublicKey,
+        secret_key: &SecretKey,
+        pass_phrase: impl AsRef<[u8]>,
+    ) -> Result<EncryptedKeys, Error> {
+        let mut rng = thread_rng();
+        let mut eraser = Eraser::new();
+        let seed = &secret_key[..SEED_LENGTH];
```

Comment:
Note that libsodium has [a dedicated function](https://download.libsodium.org/doc/public-key_cryptography/public-key_signatures#extracting-the-seed-and-the-public-key-from-the-secret-key) for seed extraction, but it seems like it's not exposed via the binding. Maybe, it could make sense to expose it?

Reply:
I guess it should be added. Have noted 👍🏻